File: scripts/showcase.py
#!/usr/bin/env -S uv run --quiet --script
# /// script
# requires-python = ">=3.11"
# dependencies = ["playwright>=1.55"]
# ///
"""Capture the README showcase images from docs/showcase/pages/*.html.

    uv run scripts/showcase.py                  # shoot every page
    uv run scripts/showcase.py loop-atlas       # shoot one
    uv run scripts/showcase.py --gif loop-atlas # animated GIF of .flow

Pages are captured from file:// — no login, no publishing, deterministic.
Full page, 2x device pixel ratio, animations frozen on their final frame
(the pages honour prefers-reduced-motion, which is what a still should show).

--gif steps the page's own CSS animations frame by frame through the Web
Animations API rather than screenshotting in real time, so the GIF is
reproducible and its timing is exactly the page's. Encoding needs ImageMagick.

The only dependency is declared above: `uv run` builds the environment on
first use and reuses it after. If Chromium is missing, uv will say so — run
`uv run --with playwright playwright install chromium` once.
"""

import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

from playwright.sync_api import sync_playwright

ROOT = Path(__file__).resolve().parent.parent
PAGES = ROOT / "docs" / "showcase" / "pages"
OUT = ROOT / "docs" / "showcase"

VIEWPORT = {"width": 1180, "height": 1000}
SCALE = 2
GIF_SELECTOR = ".flow"  # the element --gif records


def shoot(browser, name: str, frames: int = 0) -> bool:
    src = PAGES / f"{name}.html"
    if not src.exists():
        print(f"✗ no such page: {name}", file=sys.stderr)
        return False

    ctx = browser.new_context(
        viewport=VIEWPORT,
        device_scale_factor=1 if frames else SCALE,
        color_scheme="light",
        # A still of an animated page should be its final, complete frame — the
        # pages are authored so reduced-motion lands exactly there.
        reduced_motion="no-preference" if frames else "reduce",
    )
    page = ctx.new_page()
    errors: list[str] = []
    page.on("pageerror", lambda e: errors.append(str(e)))
    page.on("console", lambda m: errors.append(m.text) if m.type == "error" else None)
    page.goto(src.as_uri(), wait_until="load")
    page.evaluate("document.fonts && document.fonts.ready")

    if frames:
        target = page.query_selector(GIF_SELECTOR)
        if target is None:
            print(f"✗ {name}: no {GIF_SELECTOR} to record", file=sys.stderr)
            ctx.close()
            return False
        if not shutil.which("magick"):
            print("✗ --gif needs ImageMagick (brew install imagemagick)", file=sys.stderr)
            ctx.close()
            return False

        # Freeze the page's own animations and step them: reproducible timing.
        duration = page.evaluate("""() => {
          const anims = document.getAnimations();
          anims.forEach((a) => a.pause());
          return Math.max(0, ...anims.map((a) => {
            const t = (a.effect && a.effect.getComputedTiming) ? a.effect.getComputedTiming() : {};
            return (t.delay || 0) + (t.activeDuration || 0);
          }));
        }""")
        step = duration / frames
        with tempfile.TemporaryDirectory(prefix=f"showcase-{name}-") as tmp:
            shots = []
            for i in range(frames + 1):
                page.evaluate(
                    "(t) => document.getAnimations().forEach((a) => { a.currentTime = t; })",
                    i * step,
                )
                shot = str(Path(tmp) / f"{i:03d}.png")
                target.screenshot(path=shot)
                shots.append(shot)
            out = OUT / f"{name}.gif"
            subprocess.run(
                ["magick", "-loop", "0", "-delay", str(round(step / 10)), *shots,
                 "-delay", "220", shots[-1],            # hold on the finished picture
                 "-layers", "OptimizeTransparency", "-fuzz", "2%", str(out)],
                check=True,
            )
        print(f"✓ {name}.gif — {frames + 1} frames · {duration / 1000:.1f}s of page time")
    else:
        page.screenshot(path=str(OUT / f"{name}.png"), full_page=True)
        size = page.evaluate("""() => ({
          w: document.documentElement.scrollWidth,
          h: document.documentElement.scrollHeight,
        })""")
        print(f"✓ {name}.png — {size['w'] * SCALE}×{size['h'] * SCALE}")

    ctx.close()
    if errors:
        print(f"  ! page errors: {' | '.join(errors[:3])}", file=sys.stderr)
    return True


def main() -> int:
    args = sys.argv[1:]
    frames = 0
    if "--gif" in args:
        after = args[args.index("--gif") + 1:]
        frames = int(after[0]) if after and after[0].isdigit() else 44
    names = [a for a in args if not a.startswith("--") and not a.isdigit()]
    targets = names or sorted(p.stem for p in PAGES.glob("*.html"))

    with sync_playwright() as pw:
        browser = pw.chromium.launch()
        ok = all([shoot(browser, n, frames) for n in targets])
        browser.close()
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())

File: scripts/showcase.mjs
#!/usr/bin/env node
// Capture the README showcase images from docs/showcase/pages/*.html.
//
//   node scripts/showcase.mjs                 # shoot every page
//   node scripts/showcase.mjs loop-atlas      # shoot one
//   node scripts/showcase.mjs --gif loop-atlas      # animated GIF of .flow (needs ImageMagick)
//
// Pages are captured from file:// — no login, no publishing, deterministic.
// Full page, 2× device pixel ratio, animations frozen on their final frame
// (the pages honour prefers-reduced-motion, which is what a still should show).
//
// --gif steps the page's own CSS animations frame by frame through the Web
// Animations API rather than screenshotting in real time, so the GIF is
// reproducible and its timing is exactly the page's.
//
// Playwright is not a dependency of this repo. Resolve it from a local
// install, or from the npx cache after `npx -y playwright@1.55.0 --help`.

import { readdirSync, existsSync, mkdirSync, rmSync } from "node:fs";
import { execFileSync } from "node:child_process";
import { tmpdir } from "node:os";
import { join, dirname, basename } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";
import { homedir } from "node:os";
import { createRequire } from "node:module";

const HERE = dirname(fileURLToPath(import.meta.url));
const ROOT = join(HERE, "..");
const PAGES = join(ROOT, "docs", "showcase", "pages");
const OUT = join(ROOT, "docs", "showcase");

const VIEWPORT = { width: 1180, height: 1000 };
const GIF_SELECTOR = ".flow";   // the element --gif records
const SCALE = 2;

function loadPlaywright() {
  const req = createRequire(import.meta.url);
  const roots = [ROOT, join(homedir(), "node_modules")];
  const npx = join(homedir(), ".npm", "_npx");
  if (existsSync(npx)) {
    for (const d of readdirSync(npx)) {
      const p = join(npx, d, "node_modules");
      if (existsSync(join(p, "playwright"))) roots.push(p);
    }
  }
  for (const r of roots) {
    try { return req(join(r, "playwright")); } catch { /* next */ }
  }
  try { return req("playwright"); } catch { /* fall through */ }
  console.error(
    "playwright not found. Either:\n" +
    "  npm i -D playwright && npx playwright install chromium\n" +
    "or prime the npx cache once:\n" +
    "  npx -y playwright@1.55.0 --help");
  process.exit(1);
}

const pageFiles = () =>
  readdirSync(PAGES).filter((f) => f.endsWith(".html")).map((f) => basename(f, ".html")).sort();

async function shoot(browser, name, { frames = 0 } = {}) {
  const file = join(PAGES, `${name}.html`);
  if (!existsSync(file)) { console.error(`✗ no such page: ${name}`); return false; }

  const ctx = await browser.newContext({
    viewport: VIEWPORT,
    deviceScaleFactor: frames ? 1 : SCALE,
    colorScheme: "light",
    // A still of an animated page should be its final, complete frame — the
    // pages are authored so reduced-motion lands exactly there.
    reducedMotion: frames ? "no-preference" : "reduce",
  });
  const page = await ctx.newPage();
  const errors = [];
  page.on("pageerror", (e) => errors.push(String(e)));
  page.on("console", (m) => { if (m.type() === "error") errors.push(m.text()); });
  await page.goto(pathToFileURL(file).href, { waitUntil: "load" });
  await page.evaluate(() => document.fonts?.ready);

  if (frames) {
    const dir = join(tmpdir(), `showcase-${name}-${process.pid}`);
    mkdirSync(dir, { recursive: true });
    const target = await page.$(GIF_SELECTOR);
    if (!target) { console.error(`✗ ${name}: no ${GIF_SELECTOR} to record`); await ctx.close(); return false; }

    // Freeze the page's own animations and step them: reproducible timing.
    const duration = await page.evaluate(() => {
      const anims = document.getAnimations();
      anims.forEach((a) => a.pause());
      return Math.max(0, ...anims.map((a) => {
        const t = a.effect?.getComputedTiming?.() ?? {};
        return (t.delay ?? 0) + (t.activeDuration ?? 0);
      }));
    });
    const step = duration / frames;
    const shots = [];
    for (let i = 0; i <= frames; i++) {
      await page.evaluate((t) => document.getAnimations().forEach((a) => { a.currentTime = t; }), i * step);
      const shot = join(dir, `${String(i).padStart(3, "0")}.png`);
      await target.screenshot({ path: shot });
      shots.push(shot);
    }
    // Hold on the finished picture, then let the loop breathe before repeating.
    const out = join(OUT, `${name}.gif`);
    execFileSync("magick", [
      "-loop", "0", "-delay", String(Math.round(step / 10)), ...shots,
      "-delay", "220", shots[shots.length - 1],          // hold on the finished picture
      "-layers", "OptimizeTransparency", "-fuzz", "2%", out,
    ]);
    rmSync(dir, { recursive: true, force: true });
    console.log(`✓ ${name}.gif — ${frames + 1} frames · ${(duration / 1000).toFixed(1)}s of page time`);
  } else {
    const out = join(OUT, `${name}.png`);
    await page.screenshot({ path: out, fullPage: true });
    const { width, height } = await page.evaluate(() => ({
      width: document.documentElement.scrollWidth,
      height: document.documentElement.scrollHeight,
    }));
    console.log(`✓ ${name}.png — ${width * SCALE}×${height * SCALE}`);
  }
  await ctx.close();
  if (errors.length) console.error(`  ! page errors: ${errors.slice(0, 3).join(" | ")}`);
  return true;
}

const args = process.argv.slice(2);
const frames = args.includes("--gif") ? Number(args[args.indexOf("--gif") + 1] ?? 0) || 44 : 0;
const names = args.filter((a) => !a.startsWith("--") && !/^\d+$/.test(a));
const targets = names.length ? names : pageFiles();

const { chromium } = loadPlaywright();
const browser = await chromium.launch();
let ok = true;
for (const n of targets) ok = (await shoot(browser, n, { frames })) && ok;
await browser.close();
process.exit(ok ? 0 : 1);
